roadmap/implementers-guide/src/node/utility/availability-store.md

```diff
@@ -50,3 +93,107 @@ On `QueryChunk` message:
 On `StoreChunk` message:
 
 - Store the chunk along with its inclusion proof under the candidate hash and validator index.
+
+On `StorePoV` message:
+
+- Store the block, if the validator index is provided, store the respective chunk as well.
+
+On finality event:
+
+- For the finalized block and any earlier block (if any) update pruning records of `PoV`s and chunks to keep them for respective periods after finality.
+
+### Note any backed, included and timedout candidates in the block by `hash`.
+
+- Create a `(sender, receiver)` pair.
+- Dispatch a [`RuntimeApiMessage`][RAM]`::Request(hash, RuntimeApiRequest::CandidateEvents(sender)` and listen on the receiver for a response.
+- For every event in the response:`CandidateEvent::CandidateIncluded`.
+  * For every `CandidateEvent::CandidateBacked` do nothing
+  * For every `CandidateEvent::CandidateIncluded` update pruning records of any blocks that the node stored previously.
+  * For every `CandidateEvent::CandidateTimedOut` use pruning records to prune the data; delete the info from records.
+
+## Schema
+
+### PoV pruning
+
+We keep a record about every PoV we keep, tracking its state and the time after which this PoV should be pruned.
+
+As the state of the `Candidate` changes, so does the `Prune At` time according to the rules defined earlier.
+
+| Record 1       | .. | Record N       |
+|----------------|----|----------------|
+| CandidateHash1 | .. | CandidateHashN |
+|    Prune At    | .. |    Prune At    |
+| CandidateState | .. | CandidateState |
+
+### Chunk pruning
+
+Chunk pruning is organized in a similar schema as PoV pruning.
+
+| Record 1       | .. | Record N       |
+|----------------|----|----------------|
+| CandidateHash1 | .. | CandidateHashN |
+|    Prune At    | .. |    Prune At    |
+| CandidateState | .. | CandidateState |
+
+### Included blocks caching
+
+In order to process finality events correctly we need to cache the set of parablocks included into each relay block beginning with the last finalized block and up to the most recent heads. We have to cache this data since we are only able to query this info from the state for the `k` last blocks where `k` is a relatively small number (for more info see `Assumptions`)
+
+These are used to update Chunk pruning and PoV pruning records upon finality:
+When another block finality notification is received:
+ - For any record older than this block:
```

Comment:
Is this `O(n)` in the difference in number between the last finalized block number and the new one?

Reply:
I am worried that, especially while syncing, we may have large gaps in finality. If the node is offline for a day, or something like that. Furthermore, the races against runtime APIs might get worse in that case.

Reply:
Can we really avoid linear time here? I think that this can optimize things a bit:
1. Suppose we have to store this info on blocks `[1000..2000]`.We could do two things here:
   * shrink the data kept in this one big blob to only keeping hashes and block numbers: `[(hash1000, 1000)...(hash2000, 2000)]` and query into single records by some key like `cached_availability_info_key(hash: &Hash)`. 
   * we can also keep this blob sorted at a cheap price since we are likely to only have to append new blocks to it.
2. As soon as we get a new finality notification say about `hash1500`, we can find the exact place to chop off the recent blocks at `O(log(n))` (I guess?)
3. Chop it off, update the individual records.
4. Now we have `[(hash1501, 1501)..(hash2000, 2000)]` array.

But in any case we are going to "touch" every one of them, but hopefully only once

Reply:
in any case, i have to give it another thought

Reply:
Yeah, that is a good point. I guess what I'd want to avoid is that the subsystem gets stuck in a millions-of-blocks loop. It seems better to me to only adjust pruning some number of blocks behind the head and accidentally leak data when doing larger synchronizations. Maybe there is some room for a behind-the-scenes cleanup that does not tie up the message-processing task as well.